liblumen_alloc/src/erts/term/list.rs

```diff
@@ -367,24 +375,30 @@ impl TryFrom<TypedTerm> for Boxed<Cons> {
 }
 
 impl TryInto<String> for Boxed<Cons> {
-    type Error = Exception;
+    type Error = TryIntoStringError;
 
     fn try_into(self) -> Result<String, Self::Error> {
         self.as_ref()
             .into_iter()
             .map(|result| match result {
                 Ok(element) => {
-                    let result_char: exception::Result<char> =
-                        element.try_into().map_err(|_| badarg!().into());
+                    let result_char: Result<char, TryIntoStringError> = element
+                        .try_into()
+                        .map_err(|_| TryIntoStringError::ElementNotChar(element).into());
 
                     result_char
                 }
-                Err(_) => Err(badarg!().into()),
+                Err(_) => Err(TryIntoStringError::ImproperList),
             })
             .collect()
     }
 }
 
+pub enum TryIntoStringError {
```

Comment:
I think we can use `anyhow::Error` here as well, using either the `ImproperListError` I mentioned above, or `TypeError` in the case where an element is not valid in a charlist. We can't meaningfully use `ElementNotChar` variant anyway - if necessary, we can add the debug representation as context to the `anyhow::Error` that is returned, i.e. `error.context(format!("element is not a valid character: {:?}", term))`